### PR TITLE
feat: added anemometer(wind speed) sensor

### PR DESCRIPTION
**Pull request by QB4-dev**
This is driver for wind sensors like this one:
https://www.amazon.com/TOPINCN-Speeds-Anemometer-Environment-Anemometers/dp/B083JFLNR7

Tested on ESP8266

### CODE REVIEW

File: components/anemometer/anemometer.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file anemometer.h
+ * @defgroup misc wind
```
Comment (UncleRus) [inline components/anemometer/anemometer.h:30]: What is defined here is not the group in which the component will be placed, but an internal group, the identifier of which must be equal to the name of the component.
`@defgroup anemometer`

File: components/anemometer/anemometer.c [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file anemometer.c
+ *
+ * ESP-IDF driver for impulse wind speed sensors(anemometers)
+ *
+ * Ported from esp-open-rtos
```
Comment (UncleRus) [inline components/anemometer/anemometer.c:33]: Of course, I could be wrong, but I don’t remember such a component in esp-open-rtos...

File: components/anemometer/anemometer.c [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+{
+    anemometer_priv_t *priv = (anemometer_priv_t *)arg;
+    TickType_t ticks = xTaskGetTickCountFromISR();
+    pps_iterate(arg,ticks);
+    priv->pps++;
+}
+
+anemometer_t anemometer_init(const anemometer_config_t *conf)
+{
+	gpio_config_t io_conf;
```
Comment (UncleRus) [inline components/anemometer/anemometer.c:87]: no tabs allowed

File: components/anemometer/anemometer.c [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    /* setup GPIO */
+    io_conf.intr_type = GPIO_INTR_POSEDGE;
+    io_conf.pin_bit_mask = (1 << priv->input_pin);
+    io_conf.mode = GPIO_MODE_INPUT;
+    io_conf.pull_up_en = GPIO_PULLUP_ENABLE;
+    gpio_config(&io_conf);
+
+    /* enable interrupts */
+    gpio_install_isr_service(0);
```
Comment (UncleRus) [inline components/anemometer/anemometer.c:106]: What if the isr service is already installed in another component?

**Comment by UncleRus** [comment]
In general, I don’t really like this component that works on counting GPIO interrupts. This approach is still quite appropriate for the ESP8266, but the ESP32 has a hardware pulse counter.

If the component supported PCNT for esp32 and used GPIO interrupts as a fallback for esp8266, this would be a different matter.

In its current form, this component will not be included in the library, sorry.

File: components/anemometer/anemometer.c (hunk 2 above)
Comment (QB4-dev) [reply components/anemometer/anemometer.c]: You are right. I've copied one of the existing components as a base for new

File: components/anemometer/anemometer.c (hunk 4 above)
Comment (QB4-dev) [reply components/anemometer/anemometer.c]: added check like in wiegand driver

**Comment by QB4-dev** [comment]
> In general, I don’t really like this component that works on counting GPIO interrupts. This approach is still quite appropriate for the ESP8266, but the ESP32 has a hardware pulse counter.
> 
> If the component supported PCNT for esp32 and used GPIO interrupts as a fallback for esp8266, this would be a different matter.
> 
> In its current form, this component will not be included in the library, sorry.

You are absolutely right. When I will have access to ESP32, I'll add ESP32 hardware pulse counter. Please leave this PR as is until I add ESP32 part

**Comment by QB4-dev** [comment]
I decided to add described above changes.

Now driver is using PCNT on all ESP32 chips except ESP32C3 and on ESP-IDF v5.0.0 and above, and interrupts on ESP8266 and with older ESP-IDF

**Comment by UncleRus** [comment]
That's a horse of another color!

But please run clang-format on your code.
Then add a documentation page for your component to `/docs/source/groups`. Look at existing documentation pages to understand how to do this.
Don't forget to add a link to your documentation page in the table of contents (`/docs/source/index.rst`).

**Comment by QB4-dev** [comment]
Ok. I am gonna add documentation and of course run clang-format on my source files(sorry about that).
It may be little late but how about to rename this driver to 'impulse-sensor'?
I have written it for anemometr, but it may be useful for variety of sensor with impulse output. For example fluid flow meters, shaft rotation sensors etc.

**Comment by UncleRus** [comment]
No problem, you can rename it

**Comment by QB4-dev** [comment]
OK. Ready to review

**Comment by UncleRus** [comment]
Thank you!

**Pull request merged**